samza-test/src/main/java/org/apache/samza/test/framework/TestRunner.java

```diff
@@ -0,0 +1,343 @@
+    this.app = app;
+  }
+
+  /**
+   * Registers a system with TestRunner if not already registered and configures all the system configs to global
+   * job configs
+   */
+  private void registerSystem(String systemName) {
+    if (!systems.containsKey(systemName)) {
+      systems.put(systemName, CollectionStreamSystem.create(systemName));
+      configs.putAll(systems.get(systemName).getSystemConfigs());
+    }
+  }
+
+  /**
+   * Creates an instance of {@link TestRunner} for Low Level Samza Api
+   * @param taskClass represent a class extending either {@link StreamTask} or {@link AsyncStreamTask}
+   * @return a {@link TestRunner} for {@code taskClass}
+   */
+  public static TestRunner of(Class taskClass) {
+    Preconditions.checkNotNull(taskClass);
+    Preconditions.checkState(
+        StreamTask.class.isAssignableFrom(taskClass) || AsyncStreamTask.class.isAssignableFrom(taskClass));
+    return new TestRunner(taskClass);
+  }
+
+  /**
+   * Creates an instance of {@link TestRunner} for High Level/Fluent Samza Api
+   * @param app represent a class representing Samza job by implementing {@link StreamApplication}
+   * @return a {@link TestRunner} for {@code app}
+   */
+  public static TestRunner of(StreamApplication app) {
+    Preconditions.checkNotNull(app);
+    return new TestRunner(app);
+  }
+
+  /**
+   * Only adds a config from {@code config} to global {@code configs} if they dont exist in it.
+   * @param config represents the {@link Config} supposed to be added to global configs
+   * @return calling instance of {@link TestRunner} with added configs if they don't exist
+   */
+  public TestRunner addConfigs(Config config) {
+    Preconditions.checkNotNull(config);
+    config.forEach(this.configs::putIfAbsent);
+    return this;
+  }
+
+  /**
+   * Adds a config to {@code configs} if its not already present. Overrides a config value for which key is already
+   * exisiting in {@code configs}
+   * @param key key of the config
+   * @param value value of the config
+   * @return calling instance of {@link TestRunner} with added config
+   */
+  public TestRunner addOverrideConfig(String key, String value) {
+    Preconditions.checkNotNull(key);
+    Preconditions.checkNotNull(value);
+    configs.put(key, value);
+    return this;
+  }
+
+  /**
+   * Configures {@code stream} with the TestRunner, adds all the stream specific configs to global job configs.
+   * <p>
+   * Every stream belongs to a System (here a {@link CollectionStreamSystem}), this utility also registers the system with
+   * {@link TestRunner} if not registered already. Then it creates and initializes the stream partitions with messages for
+   * the registered System
+   * <p>
+   * @param stream represents the stream that is supposed to be configured with {@link TestRunner}
+   * @return calling instance of {@link TestRunner} with {@code stream} configured with it
+   */
+  public TestRunner addInputStream(CollectionStream stream) {
```

Comment:
In these TestRunner methods which take in a CollectionStream, could they instead just take in a more general "Stream"? In the future, you might want to have other implementations of a "Stream", so it would be good if you didn't make assumptions about using a CollectionStream. I don't think you have made TestRunner coupled too closely with CollectionStream anyways, and it is probably reasonable to leave as is now until you have another type of "Stream" to keep this simpler, but it could be worth it to keep this extension in mind so that it is not too hard to do it in the future.

Reply:
CollectionStream will be a type of StreamDescriptor (this is the more general "Stream" you are referring above will be), when we have this future api, we will replace the argument type with the base class